samza-api/src/main/java/org/apache/samza/operators/windows/TriggersBuilder.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.operators.windows;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.operators.data.MessageEnvelope;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+
+/**
+ * Programmer API for specifying {@link Triggers} for a {@link WindowFunction}.
+ *
+ * <p> A {@link Triggers} instance specifies all the early and late triggers for a {@link WindowFunction}. It also specifies how
+ * window results relate over time. Window results can be accumulating (where every result builds over its previous ones) or
+ * discarding (where every result is independent).
+ *
+ * <p> The below example windows an input into tumbling windows of 10s, and emits early results periodically every 4s in
+ * processing time, or for every 50 messages. It also specifies that window results are accumulating.
+ *
+ * <pre> {@code
+ * MessageStream<> windowedStream = stream.window(
+ *   Windows.tumblingWindow(Time.of(10, TimeUnit.SECONDS))
+ *     .setTriggers(new TriggersBuilder<>()
+ *       .withEarlyFiringsAfterCountAtleast(50)
+ *       .withEarlyFiringsAfterFirstMessage(Time.of(4, TimeUnit.SECONDS))
+ *       .accumulateFiredPanes()
+ *       .build()));
+ *    }
+ *  </pre>
+ *
+ * @param <M> type of input {@link MessageEnvelope} in the window
+ * @param <K> type of key in {@link MessageEnvelope}
+ * @param <V> type of value in the {@link MessageEnvelope}
+ *
+ */
+@InterfaceStability.Unstable
+public final class TriggersBuilder<M extends MessageEnvelope, K, V> {
```

Comment:
That sounds OK, as long as we can make sure 1) the triggers are immutable when this WindowFunction is used to create the WindowOperatorSpec; 2) Triggers created from TriggerFactory are also typed s.t. WindowFunction.setXXX() methods can have build-time type validation w/ the input Trigger object type. Let me take a look at the new code before I can give further comments on this one.